rfcs/0003-multi-blockchain-support.md

```diff
@@ -0,0 +1,534 @@
+
+A number of features in the current store interface is based on so-called
+Ethereum block pointers, a pair of block hash and block number. This is changed
+to a generic block pointer type, where the hash is a byte array rather than an
+Etheruem 256-bit hash.
+
+Another part of the current store that is specific to Ethereum is the chain
+store interface. Assuming the ongoing work on network indexing is developed far
+enough to replace the current block ingestion and block streaming, this
+interface can be removed. The only part that will have to be kept is the
+so-called chain head listener. This is moved into either the network or the
+network indexer component.
+
+### Adding Support for a New Chain
+
+This section covers what it takes to add support for a new chain. Please note
+that this RFC specifies the details only to an extend that allows to assess
+whether the direction is feasible. Names and data structure details may vary in
+the implementation.
+
+#### Graph Node
+
+Each chain integration in Graph Node lives in the `graph-node` repository under
+`chain/<name>/`, like the already existing `chain/ethereum/` support.
+
+Graph Node includes a number of new component traits and data types to enable
+multi-blockchain support:
+
+```rust
+/// A compact representation of a block on a chain.
+struct BlockPointer {
+    pub number: BigInt,
+    pub hash: Bytes,
+}
+
+/// Represents a block in a particular chain. Each chain has its own
+/// implementation of the `Block` trait.
+trait Block: ToAsc {
+    fn number(&self) -> BigInt;
+    fn hash(&self) -> Bytes;
+    fn pointer(&self) -> BlockPointer;
+    fn parent_pointer(&self) -> Option<BlockPointer>;
+}
+
+/// Represents a chain supported by the node.
+trait Chain {
+    type Network: Network;
+
+    fn new(options: ...) -> Self;
+    
+    // Looks up a network by name. Fails if the node doesn't support the network.
+    fn network(&self, name: String) -> Result<Self::Network, Error>;
+}
+
+/// Represents a network (essentially an instance of a `Chain`).
+trait Network {
+    type Block: Block;
+    type NetworkIndexer: NetworkIndexer<Self::Block>;
+    type BlockStream: BlockStream<Self::Block>;
+
+    // Methods common to all networks.
+    // These are needed by the network indexer and block stream.
+    fn latest_block(&self, options: ...) -> LatestBlockFuture;
+    fn block_by_number(&self, options: ...) -> BlockByNumberFuture;
+    fn block_by_hash(&self, options: ...) -> BlockByHashFuture;
+
+    /// Indexes network data hand handles reorgs.
+    fn indexer(&self, options: ...) -> Result<Self::NetworkIndexer, Error>;
+    
+    /// Provides a block/trigger stream for given options (including filters).
+    fn block_stream(&self, options: ...) -> Result<Self::BlockStream, Error>;
+}
+
+/// Events emitted by a network indexer.
+pub enum NetworkIndexerEvent<T: Block> {
+    ChainHead(BlockPointer),
+    Revert {
+        from: BlockPointer,
+        to: BlockPointer,
+    },
```

Comment:
Pulling revert handling out of the block stream will be nice. We probably want to make the revert algorithm analogous to the one we used in the block explorer. So this would always revert a single block.

Reply:
Yeah, outputting one `Revert` event for all blocks in a sequence of old/removed blocks is the idea. I should probably mention that in a comment above the enum variant.

Reply:
I wonder if the `from` parameter is necessary.

Reply:
I'm undecided but I'd leave it in for now. It feels useful but I'm not sure if we'd actually need it.